numpy/core/src/multiarray/ctors.c

```diff
@@ -3605,27 +3606,24 @@ array_fromfile_binary(FILE *fp, PyArray_Descr *dtype, npy_intp num, size_t *nrea
         }
         num = numbytes / dtype->elsize;
     }
-    /*
-     * When dtype->subarray is true, PyArray_NewFromDescr will decref dtype
-     * even on success, so make sure it stays around until exit.
-     */
-    Py_INCREF(dtype);
     r = (PyArrayObject *)PyArray_NewFromDescr(&PyArray_Type, dtype, 1, &num,
                                               NULL, NULL, 0, NULL);
     if (r == NULL) {
-        Py_DECREF(dtype);
         return NULL;
     }
+    /* In some cases NewFromDescr can replace the dtype, so fetch new one */
+    dtype = PyArray_DESCR(r);
```

Comment:
This I think breaks reading sub-array dtypes:
```
# After this PR
>>> import numpy as np
>>> with open('/dev/zero', 'rb') as f:
...     a = np.fromfile(f, dtype='(8,8)u8', count=1)
... 
>>> (a == 0).all()
False
```

Reply:
Oh should have thought of that. Yeah, if I do it like that, I also have to update the `num` from the actual array...